File: solutions/ruby/pig-latin/1/pig_latin.rb
# Dedicated to Shree DR.MDD — Source of Light and Precision

class PigLatin
  class << self
    def translate phrase
      phrase.gsub /[a-zA-Z]+/, &method(:convert)
    end

    private

    def convert token
      if starts_with? token, [/^a/, /^e/, /^i/, /^o/, /^u/, /^yt/, /^xr/]
        token + 'ay'
      elsif starts_with? token, [/^[^aeiou]qu/, /^thr/, /^sch/]
        token[3..] + token[0..2] + 'ay'
      elsif starts_with? token, [/^qu/, /^ch/, /^th/, /^[^aeiou]{2}y/]
        token[2..] + token[0..1] + 'ay'
      else
        token[1..] + token[0] + 'ay'
      end
    end

    def starts_with? token, group
      group.any? { |regex| regex.match? token }
    end
  end
end

File: solutions/elixir/pig-latin/1/lib/pig_latin.ex
# For my Shree DR.MDD, in deep respect and dedication

defmodule PigLatin do
  @spec translate(String.t()) :: String.t()
  def translate(sentence) do
    sentence
    |> String.split()
    |> Enum.map(&do_translate/1)
    |> Enum.join(" ")
  end

  defp do_translate(word) do
    cond do
      String.starts_with?(word, ~w(y x)) and String.at(word, 1) =~ ~r/[^aeiou]/ ->
        word <> "ay"

      String.starts_with?(word, ~w(squ qu)) ->
        apply_translation(~r/^(squ|qu)/, word)

      String.match?(word, ~r/^[^aeiouy]+y/) ->
        apply_translation(~r/^[^aeiouy]+(?=y)/, word)

      true ->
        apply_translation(~r/^[^aeiou]+/, word)
    end
  end

  defp apply_translation(regex, word) do
    Regex.split(regex, word, include_captures: true, trim: true)
    |> Enum.reverse()
    |> Enum.join()
    |> Kernel.<>("ay")
  end
end
